### PR TITLE
Close whole documents list at once

### PR DESCRIPTION
**Pull request by Richard7111**
Implemented closing of a whole document list at once, based on issue [#4760](https://github.com/apache/netbeans/issues/4760)
image

### CODE REVIEW

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (mbien) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:161]: the equals check alone would be sufficient. but to be sure, lets add a null check too. We can also use the opportunity to store the result in a variable so that we call the method only once (it could be expensive if the result is not cached).
```java
  ProjectProxy projectForTab = projectSupport.getProjectForTab(projectTab);
  if (projectForTab != null && projectForTab.equals(project)) {
```

**Review by mbien — COMMENTED** [review]
hi @Richard7111, thanks for implementing this. Ran a quick test and it worked great!

It would be good if new code would use the default NetBeans formatting, so that we don't mix so many styles in one file (no need to format the rest, only for new code). Example:
```java
    // space before braces
    private void method() {
      if (this.foo != null) {
          this.foo.method();
      }
    }
```

Please try to extract the new code which closes the tabs into a package private utility method and put it into `DocumentSwitcherTable` . Since `ButtonPupupSwitcher` is in the same package as `DocumentSwitcherTable` and can simply call its method without code duplication.

Some more comments are inline.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:160]: Yeah, when closing tab one by one, the tabIndex changes which resulted in closing only the first tab under document list. Managed to solve that by closing tab from the end

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:161]: The reason I am not checking for null, is because there is a case, when no project is associated when file and variables projectForTab and project are null. 

Should I add another `if`  for that case?

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:160]: Makes sense. Would be good to leave a comment there so that nobody changes it by accident.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:161]: good thinking. How about:
```java
  ProjectProxy projectForTab = projectSupport.getProjectForTab(projectTab);
  if ((project == null && projectForTab == null) || (projectForTab != null && projectForTab.equals(project))) {
```

Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:161]: That should work, gonna try it. Is it okay to have such a long condition?

Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:161]: sure. Its not a problem as long its somewhat readable. Your version was fine too, but without knowing about the special case where both are null, it almost looks like a bug. Making it explicit answers possible questions right away ;)

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:160]: > Please try to extract the new code which closes the tabs into a package private utility method and put it into DocumentSwitcherTable . Since ButtonPupupSwitcher is in the same package as DocumentSwitcherTable and can simply call its method without code duplication.

Somehow I pushed older version of the code, difference should be visible in the next commit. When closing tabs using DELETE key, I have to call SwingUtilities if all the tabs are closed. Otherwise the popup stays open. That is why I am keeping it outside of shared method.

Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java]: if you would move this after the loop, it should still work and we could extract the whole section into a method in `DocumentSwitcherTable`? Duplicated code is really something we should try to avoid.

Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java]: Oh yeah, you are right.

Not sure if I have done it correctly, is this the usage in ButtonPopupSwitcher?
```
DocumentSwitcherTable dst = new DocumentSwitcherTable( controller, createSwitcherItems(controller), y );
dst.closeDocumentList( item, controller );
if( controller.getTabModel().size() == 1 ) {
    SwingUtilities.invokeLater(this::hideCurrentPopup);
}
```

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (mbien) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:167]: step 1, extract this into a package-private method in the same class here.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (mbien) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:372]: step 2, remove this. and call the method:
```
pTable.closeTabsOfProject(project);
```
I think project might be the only param you need.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java]: i commented inline below. closing this thread.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:372]: Ohh, I over looked the pTable variable. thank you. Will commit asap.

Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java]: awesome

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java]: You are right. Removed parameter, but forgot to rename method. Will commit asap.

Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java]: great. Can you also as last step squash all of your commits into one and force push into this PR? Let me know if you need instructions.

**Comment by Richard7111** [comment]
Hey @mbien, I would like to write tests for my changes as a part of school assignment. However I have only used Mockito before for mocks, are you using something else for mocks or am I missing something?

**Comment by mbien** [comment]
@Richard7111 Tools like Mockito didn't exist when most of NetBeans was written, this lead to custom tooling around tests. For example most tests extend NbTestCase which can do a lot of NB specific things.

You could try to ask on the dev mailing list and see if someone has a good idea how to test things like this. However, UI tests have the tendency to break with time (sometimes randomly) and take long setup time to execute. I can't promise that we would add tests like that since we are reaching some limits. Mocking UI code is also not trivial and often not very useful.

As you can see on a full test run
https://github.com/apache/netbeans/actions/runs/3280529976/usage
 (scroll down for the total time spent), thats a lot of time spent testing.
 
Btw since we are talking about testing. I think I found a bug, I am not quite sure how to reproduce it reliably. The popup closes sometimes if a list is closed. It often happens if the first list is closed and possibly if the last list has only one item in it - had no time to investigate further.
Another thing: I believe it would be better if the popup wouldn't close if a project-name item is clicked without clicking the x just like it was before. There is no action associated with it, so I believe its better if it does nothing on click.

But there is no need to hurry to fix this since there is quite some time till the NB17 release.

**Comment by Richard7111** [comment]
@mbien hey, I just tried to replicate the bug with closing popup. It happens only when there are 2 document lists, you close first and the second one contains only one item. 

Tried to fix it, but can't seem to figure out the correct condition. Will work on it a little more.

**Comment by Richard7111** [comment]
@mbien managed to fix the bugs, feel free to check it out. Hope everything is fixed now. 😄

**Review by mbien — APPROVED** [review]
excellent. The most recent update looks good to me too.

Did some manual testing with and without project list sorting and everything worked as expected.

going to ping some more reviewers if they are interested to take a look before merging.

**Comment by mbien** [comment]
@Richard7111 I just saw that your email on the commit is now `@users.noreply.github.com`. Please fix this first to a valid address.

**Comment by eirikbakke** [comment]
Thanks for this patch!

Another thing to test: How does this work when there are tabs with unsaved changes? What if there are multiple tabs with unsaved changes?

Will take a look at the code in a bit...

**Comment by mbien** [comment]
> Another thing to test: How does this work when there are tabs with unsaved changes? What if there are multiple tabs with unsaved changes?

basically it uses the same code to close tabs as it is using now. A popup will ask the user to: discard, save or cancel

**Comment by eirikbakke** [comment]
There are some other existing bugs in ButtonPopupSwitcher to be aware of, too:
https://issues.apache.org/jira/browse/NETBEANS-5855
https://issues.apache.org/jira/browse/NETBEANS-5853

Will review...

**Comment by eirikbakke** [comment]
If multiple tabs have unsaved changes, and the user presses Cancel in the dialog box that prompts to save, what happens in this case? Probably the loop should stop in this case.

In fact, this should probably be done in the Close Other and Close All actions in the tab right-click menu as well. (CloseAllButThisAction and CloseAllDocumentsAction, respectively) But that's a separate bug.

**Comment by mbien** [comment]
> If multiple tabs have unsaved changes, and the user presses Cancel in the dialog box that prompts to save, what happens in this case? Probably the loop should stop in this case.

it just going to ask you the same question for each unedited document. The dialog has the name of the document in it.

**Comment by eirikbakke** [comment]
@mbien Right, but if the user presses Cancel (rather than No), the loop should stop.

Also, do the dialog boxes come one after another, or all on top of each other? They should probably come one at a time.

**Comment by mbien** [comment]
> @mbien Right, but if the user presses Cancel (rather than No), the loop should stop.

Not sure about this since the question is only about the concrete file. The action could simply continue with the rest.

> Also, do the dialog boxes come one after another, or all on top of each other? They should probably come one at a time.

please test it. Its a UI change which is easier to discuss if everyone played around with it. But yes they don't open all at once.

**Comment by eirikbakke** [comment]
> Not sure about this since the question is only about the concrete file. The action could simply continue with the rest.

Well, that's the difference between pressing "Cancel" and "No"; the "No" button refers to the file in question, while Cancel refers to the overall action.

I'll need to do a custom build to test this... will do but it may take a couple of days.

**Comment by mbien** [comment]
> Well, that's the difference between pressing "Cancel" and "No"; the "No" button refers to the file in question, while Cancel refers to the overall action.

It would be less intuitive to me if a file save dialog which asks me whether or not to save a single file would cancel more than what was asked in the question if cancel is pressed (it would need a "cancel all" button or something similar). The scope of the dialog seems to be clear to me: it is a save operation of a single file. The document group action is already running at this point. Some tabs might have been already closed.

To do this differently, one would have to scan the files first, ask the question with the multi-file-save dialog, then scan again and close the group if everything was saved and user didn't press cancel.

**Comment by eirikbakke** [comment]
@mbien That's standard desktop app behavior for "Cancel" in a multi-document-interface environment:  "Cancel" cancels the whole operation, "No" (actually "Discard" in NetBeans) closes the current file without saving, while continuing the loop. That's why there's a Cancel button that's separate from Discard.

> To do this differently, one would have to scan the files first, ask the question with the multi-file-save dialog, then scan again and close the group if everything was saved and user didn't press cancel.

No, that's not necessary; you just cancel when the user presses Cancel. If they already chose to save or discard some changes in other tabs, that's fine.

I just tested with VSCode and Photoshop, for instance; it works that way there. (Not applicable in IntelliJ, since it autosaves and never prompts to save the file.)

**Comment by eirikbakke** [comment]
But we can leave the Discard/Cancel distinction out of this PR... I can do a separate PR later where I also propose to adjust the similar behavior for "Close All" and "Close Other".

**Comment by Richard7111** [comment]
> @Richard7111 I just saw that your email on the commit is now `@users.noreply.github.com`. Please fix this first to a valid address.

@mbien sorry, I am not sure where you see that. I see correct mail and user name when I check `git log`. Would you mind helping me?
image

**Comment by neilcsmith-net** [comment]
IMO it would be good to align the dialog with the single one that shows when you try to close the IDE with modified files.

**Comment by mbien** [comment]
> No, that's not necessary; you just cancel when the user presses Cancel. If they already chose to save or discard some changes in other tabs, that's fine.

@eirikbakke IMO: if the situation allows it to run an atomic operation, the software should do that. Run checks first, ask if there is something to resolve, then do the operation. Don't start the op and keep asking questions while its running.

> But we can leave the Discard/Cancel distinction out of this PR... I can do a separate PR later where I also propose to adjust the similar behavior for "Close All" and "Close Other".

awesome. Would be good to have consistent behavior in comparable actions.

**Comment by eirikbakke** [comment]
@neilcsmith-net I'll look into that in a future PR.

**Comment by mbien** [comment]
> @mbien sorry, I am not sure where you see that. I see correct mail and user name when I check `git log`. Would you mind helping me a bit more?

@Richard7111 you are right! everything is fine. I looked at the wrong line:
https://github.com/apache/netbeans/actions/runs/3392742442/jobs/5645247607#step:4:16
It was the automatic merge commit, not the actual commit. All good, my mistake.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:30]: Unused import.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:313]: Unnecessary whitespace change.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:380]: Unnecessary whitespace change.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:514]: Unnecessary whitespace change.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:555]: Unnecessary whitespace change.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:357]: Should have space before "else" and after "if". Some similar formatting errors elsewhere. (Don't worry about fixing formatting issues in existing code, though.)

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:155]: What is the purpose of this condition? Under what conditions are onMouseEvent supposed to return true vs. false? (It would be good to add a Javadoc comment explaining the latter.)

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:181]: Would be good to add Javadoc explaining what the return value is supposed to mean here.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:187]: Probably better to assemble tabs to be closed in a list first, then iterate over them normally. There may be "Save Changes?" dialogs popping up as a result of trying to close the tabs, and they normally come left-to-right.

Comment (eirikbakke) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:187]: Also, there's an off-by-one error here. Reverse for loops over zero-based indices should go "for (int i = N - 1; i >= 0; i--)". (But better iterate left-to-right over a copy of the relevant tabs as suggested above.)

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [inline platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:190]: Why would the tab be closed in the project == null && projectForTab == null case?

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (eirikbakke) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:357]: Why does tabIndex == -1 indicate that closeSelectedDocumentList should be invoked? Would probably be cleaner to check if item.isSeparator(). Nice that you're handling this keypress, though!

**Review by eirikbakke — CHANGES_REQUESTED** [review]
Thanks again for this PR! Added some comments and questions.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:190]: There is a case, when no project is associated when file and variables projectForTab and project are null. Such as here for example: 
image

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:187]: I am sorry, but I dont think I understand the first suggestion.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:357]: You are right, thought I would keep the conditions similiar, but reads much better when it checks if item.isSeparator().

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:555]: Not sure what white space, it is the end of file, I have not changed anything there.  Can you give me a hint please?

Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:555]: looks like you removed the newline symbol after '}'. Its common to have one at the end. That is what the stop sign means on github.

You can fix this by putting the cursor after '}' and press return.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (mbien) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:187]: a defensive copy would look like:
`List<TabData> tabs = new ArrayList<>(controller.getTabModel().getTabs());`
instead of
`List<TabData> tabs = controller.getTabModel().getTabs();`

now you could simply iterate through `tabs` with a normal for loop since it is a copy of the original and won't change while you iterate.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/ButtonPopupSwitcher.java:555]: Oh okay, I thougth it would be that, but looked into master and did not see newline at the end. Will add it there, thanks.

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (Richard7111) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:160]: I have added Javadoc explaining return values, hope it will be clear enough.

Comment (eirikbakke) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:160]: Great, thanks!

File: platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java
Comment (eirikbakke) [reply platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java:198]: Makes sense. Thanks!

**Review by eirikbakke — APPROVED** [review]
Thanks for the changes! Tested on my own NetBeans build and confirmed to work.

**Comment by eirikbakke** [comment]
(We can probably merge this one if no one objects within a couple of days.)

**Comment by mbien** [comment]
@Richard7111 congrats to your first contribution. Thanks for your patience :)

**Comment by eirikbakke** [comment]
I added the follow-up PR relating to Close All/Close Other type actions now ( https://github.com/apache/netbeans/pull/4977 ). Also relevant since someone wanted to implement a "Close Right" action ( https://github.com/apache/netbeans/issues/4710 ).

**Pull request merged**
